bindings/haskell/test/BasicTest.hs

```diff
@@ -27,25 +28,76 @@ basicTests :: TestTree
 basicTests =
   testGroup
     "Basic Tests"
-    [ testCase "testBasicOperation" testBasicOperation
+    [ testCaseSteps "testBasicOperation" testRawOperation
+    , testCaseSteps "testMonad" testMonad
+    , testCaseSteps "testError" testError
     ]
 
-testBasicOperation :: Assertion
-testBasicOperation = do
+testRawOperation :: (String -> IO ()) -> Assertion
+testRawOperation step = do
+  step "test fs"
   Right op <- newOp "fs" $ HashMap.fromList [("root", "/tmp/opendal-test")]
-  writeOp op "key1" "value1" >>= (@?= Right ())
-  writeOp op "key2" "value2" >>= (@?= Right ())
-  readOp op "key1" >>= (@?= Right "value1")
-  readOp op "key2" >>= (@?= Right "value2")
-  isExistOp op "key1" >>= (@?= Right True)
-  isExistOp op "key2" >>= (@?= Right True)
-  createDirOp op "dir1/" >>= (@?= Right ())
-  isExistOp op "dir1/" >>= (@?= Right True)
-  copyOp op "key1" "key3" >>= (@?= Right ())
-  isExistOp op "key3" >>= (@?= Right True)
-  isExistOp op "key1" >>= (@?= Right True)
-  renameOp op "key2" "key4" >>= (@?= Right ())
-  isExistOp op "key4" >>= (@?= Right True)
-  isExistOp op "key2" >>= (@?= Right False)
-  deleteOp op "key1" >>= (@?= Right ())
-  isExistOp op "key1" >>= (@?= Right False)
\ No newline at end of file
+  writeOpRaw op "key1" "value1" >>= (@?= Right ())
+  writeOpRaw op "key2" "value2" >>= (@?= Right ())
+  readOpRaw op "key1" >>= (@?= Right "value1")
+  readOpRaw op "key2" >>= (@?= Right "value2")
+  isExistOpRaw op "key1" >>= (@?= Right True)
+  isExistOpRaw op "key2" >>= (@?= Right True)
+  createDirOpRaw op "dir1/" >>= (@?= Right ())
+  isExistOpRaw op "dir1/" >>= (@?= Right True)
+  copyOpRaw op "key1" "key3" >>= (@?= Right ())
+  isExistOpRaw op "key3" >>= (@?= Right True)
+  isExistOpRaw op "key1" >>= (@?= Right True)
+  renameOpRaw op "key2" "key4" >>= (@?= Right ())
+  isExistOpRaw op "key4" >>= (@?= Right True)
+  isExistOpRaw op "key2" >>= (@?= Right False)
+  deleteOpRaw op "key1" >>= (@?= Right ())
+  isExistOpRaw op "key1" >>= (@?= Right False)
+  step "test memory"
+  Right op2 <- newOp "memory" HashMap.empty
+  writeOpRaw op2 "key1" "value1" >>= (@?= Right ())
+  writeOpRaw op2 "key2" "value2" >>= (@?= Right ())
+  readOpRaw op2 "key1" >>= (@?= Right "value1")
+  readOpRaw op2 "key2" >>= (@?= Right "value2")
+  isExistOpRaw op2 "key1" >>= (@?= Right True)
+  isExistOpRaw op2 "key2" >>= (@?= Right True)
+  createDirOpRaw op2 "dir1/" >>= (@?= Right ())
+  isExistOpRaw op2 "dir1/" >>= (@?= Right True)
+  deleteOpRaw op2 "key1" >>= (@?= Right ())
+  isExistOpRaw op2 "key1" >>= (@?= Right False)
+
+testMonad :: (String -> IO ()) -> Assertion
+testMonad step = do
+  step "test fs"
+  Right op <- newOp "fs" $ HashMap.fromList [("root", "/tmp/opendal-test2")]
```

Comment:
Also need change.

Reply:
This has already added tests for memory

Reply:
My suggestion is to avoid testing different services. The tests should only focus on the logic of the binding itself. Additionally, it's recommended to use a cheap and easy-to-run service for testing purposes. In this case, `memory` is the best option as it has no side effects and can be run on all platforms.

Reply:
So should we only add tests for `memory`, and need to remove `fs` test?
`Memory` doesn't support `copy` and `rename` so that it can't cover all basic operations.

Reply:
> So should we only add tests for `memory`, and need to remove `fs` test?

Yes.

> `Memory` doesn't support `copy` and `rename` so that it can't cover all basic operations.

Please create an issue for this, we can add them.

Reply:
updated in new commit